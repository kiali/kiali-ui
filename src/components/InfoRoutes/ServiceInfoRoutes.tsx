import * as React from 'react';
import InfoRoutes, { Route } from './InfoRoutes';

interface Props {
  dependencies: { [key: string]: Route[] };
}

export class ServiceInfoRoutes extends React.Component<Props> {
  render() {
    return <InfoRoutes direction={'To'} resourceUrl={'workloads'} dependencies={this.props.dependencies} />;
  }
}
