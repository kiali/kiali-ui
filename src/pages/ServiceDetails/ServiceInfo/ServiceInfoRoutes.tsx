import * as React from 'react';
import { Col, Row } from 'patternfly-react';
import { Link } from 'react-router-dom';

interface ServiceInfoRoutesProps {
  dependencies?: Map<string, string[]>;
}

class ServiceInfoRoutes extends React.Component<ServiceInfoRoutesProps> {
  constructor(props: ServiceInfoRoutesProps) {
    super(props);
  }

  render() {
    return (
      <div className="card-pf">
        <Row className="row-cards-pf">
          <Col xs={12} sm={12} md={12} lg={12}>
            {Object.keys(this.props.dependencies || new Map()).map((key, u) => (
              <div className="card-pf-body" key={'dependencies_' + u}>
                <div className="progress-description">
                  <strong>To: </strong> {key}
                </div>
                <ul style={{ listStyleType: 'none' }}>
                  {(this.props.dependencies ? this.props.dependencies[key] : []).map((dependency, i) => {
                    let nVersion = dependency.indexOf('/');
                    let nNamespace = dependency.indexOf('.');
                    let servicename = dependency.substring(0, nNamespace);
                    let namespace = dependency.substring(nNamespace + 1, nVersion);
                    if (servicename.length > 0 && namespace.length > 0) {
                      let to = '/namespaces/' + namespace + '/services/' + servicename;
                      return (
                        <Link key={to + key + dependency} to={to}>
                          <li key={'dependencies_' + u + '_dependency_' + i}>{dependency}</li>
                        </Link>
                      );
                    } else {
                      return <li key={'dependencies_' + u + '_dependency_' + i}>{dependency}</li>;
                    }
                  })}
                </ul>
              </div>
            ))}
          </Col>
        </Row>
      </div>
    );
  }
}

export default ServiceInfoRoutes;
