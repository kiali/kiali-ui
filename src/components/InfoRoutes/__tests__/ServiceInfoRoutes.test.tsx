import * as React from 'react';
import { render } from 'enzyme';
import { ServiceInfoRoutes } from '../ServiceInfoRoutes';
import { BrowserRouter as Router } from 'react-router-dom';

let wrapper;

describe('#ServiceInfoRoutes render correctly with data', () => {
  beforeEach(() => {
    const dependencies = {
      v1: [{ namespace: 'bookinfo', name: 'reviews-v1' }, { namespace: 'istio-system', name: 'istio-citadel' }]
    };

    wrapper = render(
      <Router>
        <div>
          <ServiceInfoRoutes dependencies={dependencies} />
        </div>
      </Router>
    );
  });

  it('renders', () => {
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  it('shows properly the from header', () => {
    const destinationElem = wrapper.find('.progress-description > strong');
    expect(destinationElem[0].children[0].data).toEqual('To: ');
    expect(destinationElem[0].next.data).toEqual(' v1');
  });

  it('renders service routes', () => {
    const expectationData = {
      'reviews-v1': { href: '/namespaces/bookinfo/workloads/reviews-v1' },
      'istio-citadel': { href: '/namespaces/istio-system/workloads/istio-citadel' }
    };

    wrapper.find('a').each((index, selector) => {
      const serviceName = selector.children[0].data;
      expect(serviceName).toBeDefined();
      if (!serviceName) {
        return;
      }
      expect(selector.attribs['href']).toEqual(expectationData[serviceName]['href']);
    });
  });
});
