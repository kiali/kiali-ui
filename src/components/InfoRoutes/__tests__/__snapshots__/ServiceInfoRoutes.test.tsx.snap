// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#ServiceInfoRoutes render correctly with data renders 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "card-pf",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "row-cards-pf row",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "col-lg-12 col-md-12 col-sm-12 col-xs-12",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "class": "card-pf-body",
                    },
                    "children": Array [
                      Object {
                        "attribs": Object {
                          "class": "progress-description",
                        },
                        "children": Array [
                          Object {
                            "attribs": Object {},
                            "children": Array [
                              Object {
                                "data": "To: ",
                                "next": null,
                                "parent": [Circular],
                                "prev": null,
                                "type": "text",
                              },
                            ],
                            "name": "strong",
                            "namespace": "http://www.w3.org/1999/xhtml",
                            "next": Object {
                              "data": " v1",
                              "next": null,
                              "parent": [Circular],
                              "prev": [Circular],
                              "type": "text",
                            },
                            "parent": [Circular],
                            "prev": null,
                            "type": "tag",
                            "x-attribsNamespace": Object {},
                            "x-attribsPrefix": Object {},
                          },
                          Object {
                            "data": " v1",
                            "next": null,
                            "parent": [Circular],
                            "prev": Object {
                              "attribs": Object {},
                              "children": Array [
                                Object {
                                  "data": "To: ",
                                  "next": null,
                                  "parent": [Circular],
                                  "prev": null,
                                  "type": "text",
                                },
                              ],
                              "name": "strong",
                              "namespace": "http://www.w3.org/1999/xhtml",
                              "next": [Circular],
                              "parent": [Circular],
                              "prev": null,
                              "type": "tag",
                              "x-attribsNamespace": Object {},
                              "x-attribsPrefix": Object {},
                            },
                            "type": "text",
                          },
                        ],
                        "name": "div",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": Object {
                          "attribs": Object {
                            "style": "list-style-type:none",
                          },
                          "children": Array [
                            Object {
                              "attribs": Object {},
                              "children": Array [
                                Object {
                                  "attribs": Object {
                                    "href": "/namespaces/bookinfo/workloads/reviews-v1",
                                  },
                                  "children": Array [
                                    Object {
                                      "data": "reviews-v1",
                                      "next": null,
                                      "parent": [Circular],
                                      "prev": null,
                                      "type": "text",
                                    },
                                  ],
                                  "name": "a",
                                  "namespace": "http://www.w3.org/1999/xhtml",
                                  "next": null,
                                  "parent": [Circular],
                                  "prev": null,
                                  "type": "tag",
                                  "x-attribsNamespace": Object {
                                    "href": undefined,
                                  },
                                  "x-attribsPrefix": Object {
                                    "href": undefined,
                                  },
                                },
                              ],
                              "name": "li",
                              "namespace": "http://www.w3.org/1999/xhtml",
                              "next": Object {
                                "attribs": Object {},
                                "children": Array [
                                  Object {
                                    "attribs": Object {
                                      "href": "/namespaces/istio-system/workloads/istio-citadel",
                                    },
                                    "children": Array [
                                      Object {
                                        "data": "istio-citadel",
                                        "next": null,
                                        "parent": [Circular],
                                        "prev": null,
                                        "type": "text",
                                      },
                                    ],
                                    "name": "a",
                                    "namespace": "http://www.w3.org/1999/xhtml",
                                    "next": null,
                                    "parent": [Circular],
                                    "prev": null,
                                    "type": "tag",
                                    "x-attribsNamespace": Object {
                                      "href": undefined,
                                    },
                                    "x-attribsPrefix": Object {
                                      "href": undefined,
                                    },
                                  },
                                ],
                                "name": "li",
                                "namespace": "http://www.w3.org/1999/xhtml",
                                "next": null,
                                "parent": [Circular],
                                "prev": [Circular],
                                "type": "tag",
                                "x-attribsNamespace": Object {},
                                "x-attribsPrefix": Object {},
                              },
                              "parent": [Circular],
                              "prev": null,
                              "type": "tag",
                              "x-attribsNamespace": Object {},
                              "x-attribsPrefix": Object {},
                            },
                            Object {
                              "attribs": Object {},
                              "children": Array [
                                Object {
                                  "attribs": Object {
                                    "href": "/namespaces/istio-system/workloads/istio-citadel",
                                  },
                                  "children": Array [
                                    Object {
                                      "data": "istio-citadel",
                                      "next": null,
                                      "parent": [Circular],
                                      "prev": null,
                                      "type": "text",
                                    },
                                  ],
                                  "name": "a",
                                  "namespace": "http://www.w3.org/1999/xhtml",
                                  "next": null,
                                  "parent": [Circular],
                                  "prev": null,
                                  "type": "tag",
                                  "x-attribsNamespace": Object {
                                    "href": undefined,
                                  },
                                  "x-attribsPrefix": Object {
                                    "href": undefined,
                                  },
                                },
                              ],
                              "name": "li",
                              "namespace": "http://www.w3.org/1999/xhtml",
                              "next": null,
                              "parent": [Circular],
                              "prev": Object {
                                "attribs": Object {},
                                "children": Array [
                                  Object {
                                    "attribs": Object {
                                      "href": "/namespaces/bookinfo/workloads/reviews-v1",
                                    },
                                    "children": Array [
                                      Object {
                                        "data": "reviews-v1",
                                        "next": null,
                                        "parent": [Circular],
                                        "prev": null,
                                        "type": "text",
                                      },
                                    ],
                                    "name": "a",
                                    "namespace": "http://www.w3.org/1999/xhtml",
                                    "next": null,
                                    "parent": [Circular],
                                    "prev": null,
                                    "type": "tag",
                                    "x-attribsNamespace": Object {
                                      "href": undefined,
                                    },
                                    "x-attribsPrefix": Object {
                                      "href": undefined,
                                    },
                                  },
                                ],
                                "name": "li",
                                "namespace": "http://www.w3.org/1999/xhtml",
                                "next": [Circular],
                                "parent": [Circular],
                                "prev": null,
                                "type": "tag",
                                "x-attribsNamespace": Object {},
                                "x-attribsPrefix": Object {},
                              },
                              "type": "tag",
                              "x-attribsNamespace": Object {},
                              "x-attribsPrefix": Object {},
                            },
                          ],
                          "name": "ul",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": [Circular],
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "style": undefined,
                          },
                          "x-attribsPrefix": Object {
                            "style": undefined,
                          },
                        },
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {
                          "class": undefined,
                        },
                        "x-attribsPrefix": Object {
                          "class": undefined,
                        },
                      },
                      Object {
                        "attribs": Object {
                          "style": "list-style-type:none",
                        },
                        "children": Array [
                          Object {
                            "attribs": Object {},
                            "children": Array [
                              Object {
                                "attribs": Object {
                                  "href": "/namespaces/bookinfo/workloads/reviews-v1",
                                },
                                "children": Array [
                                  Object {
                                    "data": "reviews-v1",
                                    "next": null,
                                    "parent": [Circular],
                                    "prev": null,
                                    "type": "text",
                                  },
                                ],
                                "name": "a",
                                "namespace": "http://www.w3.org/1999/xhtml",
                                "next": null,
                                "parent": [Circular],
                                "prev": null,
                                "type": "tag",
                                "x-attribsNamespace": Object {
                                  "href": undefined,
                                },
                                "x-attribsPrefix": Object {
                                  "href": undefined,
                                },
                              },
                            ],
                            "name": "li",
                            "namespace": "http://www.w3.org/1999/xhtml",
                            "next": Object {
                              "attribs": Object {},
                              "children": Array [
                                Object {
                                  "attribs": Object {
                                    "href": "/namespaces/istio-system/workloads/istio-citadel",
                                  },
                                  "children": Array [
                                    Object {
                                      "data": "istio-citadel",
                                      "next": null,
                                      "parent": [Circular],
                                      "prev": null,
                                      "type": "text",
                                    },
                                  ],
                                  "name": "a",
                                  "namespace": "http://www.w3.org/1999/xhtml",
                                  "next": null,
                                  "parent": [Circular],
                                  "prev": null,
                                  "type": "tag",
                                  "x-attribsNamespace": Object {
                                    "href": undefined,
                                  },
                                  "x-attribsPrefix": Object {
                                    "href": undefined,
                                  },
                                },
                              ],
                              "name": "li",
                              "namespace": "http://www.w3.org/1999/xhtml",
                              "next": null,
                              "parent": [Circular],
                              "prev": [Circular],
                              "type": "tag",
                              "x-attribsNamespace": Object {},
                              "x-attribsPrefix": Object {},
                            },
                            "parent": [Circular],
                            "prev": null,
                            "type": "tag",
                            "x-attribsNamespace": Object {},
                            "x-attribsPrefix": Object {},
                          },
                          Object {
                            "attribs": Object {},
                            "children": Array [
                              Object {
                                "attribs": Object {
                                  "href": "/namespaces/istio-system/workloads/istio-citadel",
                                },
                                "children": Array [
                                  Object {
                                    "data": "istio-citadel",
                                    "next": null,
                                    "parent": [Circular],
                                    "prev": null,
                                    "type": "text",
                                  },
                                ],
                                "name": "a",
                                "namespace": "http://www.w3.org/1999/xhtml",
                                "next": null,
                                "parent": [Circular],
                                "prev": null,
                                "type": "tag",
                                "x-attribsNamespace": Object {
                                  "href": undefined,
                                },
                                "x-attribsPrefix": Object {
                                  "href": undefined,
                                },
                              },
                            ],
                            "name": "li",
                            "namespace": "http://www.w3.org/1999/xhtml",
                            "next": null,
                            "parent": [Circular],
                            "prev": Object {
                              "attribs": Object {},
                              "children": Array [
                                Object {
                                  "attribs": Object {
                                    "href": "/namespaces/bookinfo/workloads/reviews-v1",
                                  },
                                  "children": Array [
                                    Object {
                                      "data": "reviews-v1",
                                      "next": null,
                                      "parent": [Circular],
                                      "prev": null,
                                      "type": "text",
                                    },
                                  ],
                                  "name": "a",
                                  "namespace": "http://www.w3.org/1999/xhtml",
                                  "next": null,
                                  "parent": [Circular],
                                  "prev": null,
                                  "type": "tag",
                                  "x-attribsNamespace": Object {
                                    "href": undefined,
                                  },
                                  "x-attribsPrefix": Object {
                                    "href": undefined,
                                  },
                                },
                              ],
                              "name": "li",
                              "namespace": "http://www.w3.org/1999/xhtml",
                              "next": [Circular],
                              "parent": [Circular],
                              "prev": null,
                              "type": "tag",
                              "x-attribsNamespace": Object {},
                              "x-attribsPrefix": Object {},
                            },
                            "type": "tag",
                            "x-attribsNamespace": Object {},
                            "x-attribsPrefix": Object {},
                          },
                        ],
                        "name": "ul",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": null,
                        "parent": [Circular],
                        "prev": Object {
                          "attribs": Object {
                            "class": "progress-description",
                          },
                          "children": Array [
                            Object {
                              "attribs": Object {},
                              "children": Array [
                                Object {
                                  "data": "To: ",
                                  "next": null,
                                  "parent": [Circular],
                                  "prev": null,
                                  "type": "text",
                                },
                              ],
                              "name": "strong",
                              "namespace": "http://www.w3.org/1999/xhtml",
                              "next": Object {
                                "data": " v1",
                                "next": null,
                                "parent": [Circular],
                                "prev": [Circular],
                                "type": "text",
                              },
                              "parent": [Circular],
                              "prev": null,
                              "type": "tag",
                              "x-attribsNamespace": Object {},
                              "x-attribsPrefix": Object {},
                            },
                            Object {
                              "data": " v1",
                              "next": null,
                              "parent": [Circular],
                              "prev": Object {
                                "attribs": Object {},
                                "children": Array [
                                  Object {
                                    "data": "To: ",
                                    "next": null,
                                    "parent": [Circular],
                                    "prev": null,
                                    "type": "text",
                                  },
                                ],
                                "name": "strong",
                                "namespace": "http://www.w3.org/1999/xhtml",
                                "next": [Circular],
                                "parent": [Circular],
                                "prev": null,
                                "type": "tag",
                                "x-attribsNamespace": Object {},
                                "x-attribsPrefix": Object {},
                              },
                              "type": "text",
                            },
                          ],
                          "name": "div",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": [Circular],
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "class": undefined,
                          },
                          "x-attribsPrefix": Object {
                            "class": undefined,
                          },
                        },
                        "type": "tag",
                        "x-attribsNamespace": Object {
                          "style": undefined,
                        },
                        "x-attribsPrefix": Object {
                          "style": undefined,
                        },
                      },
                    ],
                    "name": "div",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
